app/models/concerns/has_custom_fields.rb

```diff
@@ -86,6 +86,10 @@ def append_custom_field(target, key, value)
     end
 
     def register_custom_field_type(name, type)
+      if Array === type
+        Discourse.deprecate("Array types for custom fields are deprecated, use type :json instead")
```

Comment:
Can we add a `drop_from` parameter? I suspect a 3.3.0 version might be reasonable (that release would be scheduled for July/August 2024 at our current cadence).